hub/core/tensor.py

```diff
@@ -0,0 +1,170 @@
+    """A tensor exists if at the specified `key` and `storage` there is both a tensor meta file and index map."""
+
+    meta_key = get_tensor_meta_key(key)
+    index_map_key = get_index_map_key(key)
+    return meta_key in storage and index_map_key in storage
+
+
+def create_tensor(key: str, storage: StorageProvider, meta: dict):
+    """If a tensor does not exist, create a new one with the provided meta.
+
+    Args:
+        key (str): Key for where the chunks, index_map, and meta will be located in `storage` relative to it's root.
+        storage (StorageProvider): StorageProvider that all tensor data is written to.
+        meta (dict): Meta for the tensor. Required Properties:
+            chunk_size (int): Desired length of chunks.
+            dtype (str): Datatype for each sample.
+
+    Raises:
+        TensorAlreadyExistsError: If a tensor defined with `key` already exists.
+    """
+
+    if tensor_exists(key, storage):
+        raise TensorAlreadyExistsError(key)
+
+    meta.update({"length": 0})
+
+    validate_tensor_meta(meta)
+
+    write_tensor_meta(key, storage, meta)
+    write_index_map(key, storage, [])
+
+
+def add_samples_to_tensor(
+    array: np.ndarray,
+    key: str,
+    storage: StorageProvider,
+    batched: bool = False,
+):
+    """Create a new tensor (if one doesn't already exist), then chunk and write the given array to storage.
+
+    For more on chunking, see the `generate_chunks` method.
+
+    Args:
+        array (np.ndarray): Array to be chunked/written. Batch axis (`array.shape[0]`) is optional, if `array` does have a
+            batch axis, you should pass the argument `batched=True`.
+        key (str): Key for where the chunks, index_map, and meta will be located in `storage` relative to it's root.
+        storage (StorageProvider): StorageProvider for storing the chunks, index_map, and meta.
+        batched (bool): If True, the provied `array`'s first axis (`shape[0]`) will be considered it's batch axis.
+            If False, a new axis will be created with a size of 1 (`array.shape[0] == 1`). default=False
+
+    raises:
+        TensorDoesNotExistError: If a tensor at `key` does not exist. A tensor must be created first using `create_tensor(...)`.
+    """
+
+    array = normalize_and_batchify_shape(array, batched=batched)
+
+    if not tensor_exists(key, storage):
+        raise TensorDoesNotExistError(key)
+
+    index_map = read_index_map(key, storage)
+    tensor_meta = read_tensor_meta(key, storage)
+    _check_array_and_tensor_are_compatible(tensor_meta, array)
+
+    # TODO: get the tobytes function from meta
+    tobytes = row_wise_to_bytes
+
+    array_length = array.shape[0]
+    for i in range(array_length):
+        sample = array[i]
+
+        # TODO: we may want to call `tobytes` on `array` and call memoryview on that. this may depend on the access patterns we
+        # choose to optimize for.
+        b = memoryview(tobytes(sample))
+
+        index_map_entry = write_bytes(
```

Comment:
nit-pick: Add a TODO to improve this, We call "get_last_chunk" for each sample and that can be memoized and reused.